arch/riscv/kernel/devicetree.c

```diff
@@ -0,0 +1,62 @@
+/*
+ * Enables use of Device Tree for RISC-V peripheral devices.
+ *
+ * Author: Jamey Hicks <jamey.hicks@gmail.com>
+ */
+
+#include <linux/module.h>
+#include <linux/types.h>
+#include <linux/kernel.h>
+#include <linux/init.h>
+#include <linux/slab.h>
+#include <linux/device.h>
+#include <linux/io.h>
+#include <linux/irqchip.h>
+#include <linux/of_fdt.h>
+#include <linux/of_platform.h>
+
+int __init devicetree_init(void)
+{
+	resource_size_t initial_dtb = 0x04100000;
```

Comment:
>  I think the boot loader should be able to pass in the address of the device tree, either on the command line or via a register. What do you recommend?

According to `Documentation/devicetree/usage-model.txt`, the kernel allows the command line string to be passed via the `bootargs` property, so using a boot parameter to convey the device tree address introduces a potential chicken-and-egg problem.

The usual practice is to pass by register (see `Documentation/devicetree/booting-without-of.txt`), but to avoid confusion, I would prefer to wait until a bootloader interface is standardized as part of the platform specification, now being actively developed by others.  In the meantime, adding a temporary Kconfig knob might be the cleanest option, for which there is at least one precedent: `CONFIG_NIOS2_DTB_AT_PHYS_ADDR` in the nios2 port.